### ISSUE TITLE
F5 Devices - SSH KEEPALIVE ISSUE

### ISSUE THREAD

**Issue by MikeJurecki** [ktbyers/netmiko#692]
I am having an issue where I am pulling in a list of F5 devices and then checking TACACS logins and one other check, but it keeps hanging after it runs through about 20 devices.  Basically it logs in using my user creds, runs a command, disconnects and then logs the same device with a PWoLR login and runs the same command and finally disconnects and moves onto the next device.  

The issue is that after it runs through roughly 20 devices it gets stuck sending the following keepalive packets, see log entry below:
"**DEBUG:paramiko.transport:[chan 0] Unhandled channel request "keepalive@openssh.com**"

It just keeps sending them and won't timeout.  If I remove the device from the host list and rerun the script it will do it on another device at about 20 devices in again.  

I am trying to think of a way to have it close the connection if it sends out "x" amount of keepalive packets.  Is there a way to do this or any way to have it timeout even if it didnt finish running after "y" amount of seconds?

Here is my code:

**Note**: I only posted at the start of the Network Connection Code (There is a large chuck more above as I am utilizing xlsxwriter for output, let me know if you want that too)

```
# LIBRARIES USED
import netmiko
import xlsxwriter
import logging
import time
from datetime import datetime
from development import F5_email

# START OF NETWORK SECTION

creds = []  #USER CREDS
with open(r'C:\Users\user\PycharmProjects\PNM_F5_Automation\development\creds\creds3.txt', 'r') as file:
    creds = [line.strip() for line in file]

username = creds[0]
password = creds[1]

creds2 = []  # PWoLR CREDS
with open(r'C:\Users\user\PycharmProjects\PNM_F5_Automation\development\creds\creds2.txt', 'r') as file:
    creds2 = [line.strip() for line in file]

username2 = creds2[0]
password2 = creds2[1]

iostype = 'f5_ltm'
hostfile = open(r'C:\Users\user\PycharmProjects\PNM_F5_Automation\development\hostlist.txt', 'r')

row = 1

# USER CREDS CHECK START

for ipaddress in hostfile:
    ipaddress = ipaddress.strip()

    try:
        net_connect = netmiko.ConnectHandler(device_type=iostype, ip=ipaddress, username=username,
                                             password=password, port=22, secret=password, verbose=False)

        out_worksheet1.write(row, 0, ipaddress, normal_format2)

        syscheck = net_connect.send_command('list sys syslog', delay_factor=2)
        out_worksheet1.write(row, 2, syscheck, normal_format)

        out_worksheet1.write(row, 1, '(USER) TACACS Check Successful', passed_format)

        net_connect.disconnect()

    except netmiko.NetMikoAuthenticationException:
        print('(USER) Invalid Credentials: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 0, ipaddress, failed_format)
        out_worksheet1.write(row, 1, '(USER) Invalid Credentials: Failed Check', failed_format)
        pass


    except netmiko.NetMikoTimeoutException:
        print('(USER) SSH Error: Connection Failed: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 0, ipaddress, failed_format)
        out_worksheet1.write(row, 1, '(USER) SSH Connection Error', failed_format)
        pass


    except:
        print('(USER) Connection Failed For Unknown Reason: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 0, ipaddress, failed_format)
        out_worksheet1.write(row, 1, '(USER) Connection Failed: Unknown Reason Reference Log', failed_format)
        pass

    else:
        print('(USER) Connection to', ipaddress, 'established:', datetime.now().time())

    # END OF USER CHECK
    time.sleep(1)
    # START OF PWoLR CHECK

    try:
        net_connect2 = netmiko.ConnectHandler(device_type=iostype, ip=ipaddress, username=username2,
                                             password=password2, port=22, secret=password2, verbose=False)

        syscheck = net_connect2.send_command('list sys syslog', delay_factor=2)
        out_worksheet1.write(row, 4, syscheck, normal_format)

        out_worksheet1.write(row, 3, '(PWoLR) TACACS Check Successful', passed_format)

        net_connect2.disconnect()

    except netmiko.NetMikoAuthenticationException:
        print('(PWoLR) Invalid Credentials: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 3, '(PWoLR) Invalid Credentials: Failed Check', failed_format)
        pass


    except netmiko.NetMikoTimeoutException:
        print('(PWoLR) SSH Error: Connection Failed: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 3, '(PWoLR) SSH Connection Error', failed_format)
        pass

    except:
        print('(PWoLR) Connection Failed For Unknown Reason: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 3, '(PWoLR) Connection Failed: Unknown Reason Reference Log', failed_format)
        pass

    else:
        print('(PWoLR) Connection to', ipaddress, 'established:', datetime.now().time())

    row += 1
    time.sleep(1)

out_workbook.close()

time.sleep(1)

F5_email.send_email(attachments=[o_filename])

#END OF SCRIPT
```

**Comment by ktbyers** [comment]
Can you turn on logging and post the log file here (obscure any confidential information like passwords or IP addresses):

https://github.com/ktbyers/netmiko/blob/develop/COMMON_ISSUES.md#enable-netmiko-logging-of-all-reads-and-writes-of-the-communications-channel


It looks like the F5 is possibly doing something unusual from Paramiko's perspective. Do you see anything in the device logs or in the TACACS logs?.

**Comment by BigPat** [comment]
So the issue was that the device had an extremely slow response time but when I let it sit eventually it ran.  
--------------------------------------------------------------------------------------------------------
However, now I am getting an issue where after it runs the first command on the F5 it just gets stuck.  I think it is due to implementing an is_alive function.  I tried adding in a command to send "q" but that doesnt seem to fix it so I removed it.  Unless my pycharm is messing up I do not understand why it wont send the command.  

```
import socket
import netmiko
import xlsxwriter
import logging
import time
from datetime import datetime

# Cut out start of code

def is_alive(sshtest):
    null = chr(0)

    try:
        # Try sending ASCII null byte to maintain the connection alive
        sshtest.write_channel(null)
        return True
    except (socket.error, EOFError):
        return False

for ipaddress in hostfile:
    ipaddress = ipaddress.strip()

    # --------------------------------------------------------------------------------------------------------------
    #                                            START OF CRED CHECK
    # --------------------------------------------------------------------------------------------------------------
    try:
        net_connect = netmiko.ConnectHandler(device_type=iostype, ip=ipaddress, username=username,
                                             password=password, port=22, secret=password, verbose=False)

        out_worksheet1.write(row, 1, ipaddress, normal_format2)

        if is_alive(sshtest=net_connect) is True:
            print('Connection to', ipaddress, 'established:', datetime.now().time())
            out_worksheet1.write(row, 2, 'Passed', passed_format)

            try:
                syscheck = net_connect.send_command('list sys syslog', delay_factor=2) # Runs this command
                out_worksheet1.write(row, 3, syscheck, normal_format)
                # TODO FIX THIS, Next Line will not run.
                hostN = net_connect.send_command('list sys global-settings hostname', delay_factor=2) # Will not run
                hostN = hostN.splitlines()
                hostN2 = hostN[1]
                hostN2 = hostN2[13:]
                hostN2 = hostN2.strip()
                out_worksheet1.write(row, 0, hostN2, normal_format2)

            except:
                print('command failed')
                pass

        else:
            print('Connection is not alive')
            net_connect.disconnect()

    except netmiko.NetMikoAuthenticationException:
        print('Invalid Credentials: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 1, ipaddress, failed_format)
        out_worksheet1.write(row, 2, 'Failed', failed_format)
        pass

    except netmiko.NetMikoTimeoutException:
        print('SSH Error: Connection Failed: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 1, ipaddress, failed_format)
        out_worksheet1.write(row, 2, 'Failed', failed_format)
        pass

    except:
        print('Connection Failed For Unknown Reason: ', ipaddress)
        print('--- ')
        out_worksheet1.write(row, 1, ipaddress, failed_format)
        out_worksheet1.write(row, 2, 'Failed', failed_format)
        pass

    print('Made it here: END1')
    time.sleep(1)


```


In the log it makes it to the 'list sys log" command and then it just does nothing...It did not do this before which is odd.  

```
user@hostname(Active)(tmos)# 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:write_channel: b'\x00'
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:write_channel: b'\n'
DEBUG:netmiko:read_channel: ^J
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:write_channel: b'list sys syslog\n'
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: list sys syslog

sys syslog {

    remote-servers {

        custom-app {

            host 0.0.0.0

            remote-port 5555555

        }

    }

}

user@hostname(Active)(tmos)# 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 
DEBUG:netmiko:read_channel: 

```

In the log it just keeps going and "DEBUG:netmiko:read_channel: " messages keep occuring.  

Basically this script is meant to login with domain creds, logout and then login with PWoLR to ensure every device is configured correctly for TACACS and local admin.  My issue is that if the first one fails then it will just post "success" in my output file if I take the statement out of the try statement, which is why I added in the is_alive() function.  

Any suggestions or advice would be greatly appreciated.

**Comment by BigPat** [comment]
Note that I also tried taking out the first command and then the second command worked just fine but it got stuck again and wouldn't disconnect from the device.  


```

            try:
                #syscheck = net_connect.send_command('list sys syslog', delay_factor=2) # Runs this command
                #out_worksheet1.write(row, 3, syscheck, normal_format)
                # TODO FIX THIS, Next Line will not run.
                hostN = net_connect.send_command('list sys global-settings hostname', delay_factor=2) # Will not run
                hostN = hostN.splitlines()
                hostN2 = hostN[1]
                hostN2 = hostN2[13:]
                hostN2 = hostN2.strip()
                out_worksheet1.write(row, 0, hostN2, normal_format2)
                net_connect.disconnect()

            except:
                print('command failed')
                pass
```
